### ISSUE TITLE
Improvement for signing with PFX files & WPF problems

### ISSUE THREAD

**Issue by lextudio-support** [obfuscar/obfuscar#525]
Hi there! :)
Congratulations for your Project! :)
I am trying to obfuscate a WPF app and apart from the problems that I will mention at the end, I have an improvement that will allow signing with PFX files. Here is the improvement:

Project.cs, Line 118 (replaced old code with this and added a "PfxPsw" element in the configuration XML)

```

				} else {
					try
					{
						var keyFile = vars.GetValue("KeyFile", null);
						var bytes = File.ReadAllBytes(keyFile);
						var ext = Path.GetExtension(keyFile);
						if (ext!=null && ext.ToLower() == ".pfx")
						{
							bytes = Pfx2Snk(bytes, vars.GetValue("PfxPsw", null));
						}
						else
						{
						}
						keyvalue = CryptoConvert.FromCapiKeyBlob(bytes);
					} catch (Exception ex) {
						throw new ObfuscarException (String.Format ("Failure loading key file \"{0}\"", vars.GetValue ("KeyFile", null)), ex);
					}
```
```
		/// <summary>
		/// Converts .pfx file to .snk file.
		/// </summary>
		/// <param name="pfxData">.pfx file data.</param>
		/// <param name="pfxPassword">.pfx file password.</param>
		/// <returns>.snk file data.</returns>
		public static byte[] Pfx2Snk(byte[] pfxData, string pfxPassword)
		{
			// load .pfx
			var cert = new X509Certificate2(pfxData, pfxPassword, X509KeyStorageFlags.Exportable);

			// create .snk
			var privateKey = (RSACryptoServiceProvider)cert.PrivateKey;
			return privateKey.ExportCspBlob(true);
		}
```

Now to the WPF problems:
I use this XML configuration that should not change the assemblies at all:

```
<?xml version='1.0'?>
<Obfuscator>
  <Var name="InPath" value="C:\Users\Me\" />
  <Var name="OutPath" value="$(InPath)\Obfuscar_Output" />

  <Var name="MarkedOnly" value="true" />
  <Var name="KeepPublicApi" value="true" />
  <Var name="HidePrivateApi" value="false" />
  <Var name="ReuseNames" value="false" />
  <Var name="HideStrings" value="false" />
  <Var name="OptimizedMethods" value="false" />
  <Var name="RenameProperties" value="false" />
  <Var name="RenameEvents" value="false" />
  <Var name="XmlMapping" value="true" />
  <Var name="PfxPsw" value="myPfxPassword" />
  <Var name="KeyFile" value=".\Key.pfx" />

  <Module file="$(InPath)\Dll1.dll" />
  <Module file="$(InPath)\Dll2.dll" />
  <Module file="$(InPath)\EXE1.exe" />
  <Module file="$(InPath)\EXE2.exe" />
</Obfuscator>
```
However, when i start the WPF app, it crashes and in the event log I see this:

> Exception Info: System.Windows.Markup.XamlParseException
Stack:
   at System.Windows.Markup.WpfXamlLoader.Load(System.Xaml.XamlReader, System.Xaml.IXamlObjectWriterFactory, Boolean, System.Object, System.Xaml.XamlObjectWriterSettings, System.Uri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(System.Xaml.XamlReader, Boolean, System.Object, System.Xaml.Permissions.XamlAccessLevel, System.Uri)
   at System.Windows.Markup.XamlReader.LoadBaml(System.IO.Stream, System.Windows.Markup.ParserContext, System.Object, Boolean)
   at System.Windows.Application.LoadComponent(System.Object, System.Uri)
   at KEL_Building_Manager.App.Main()

**Original Reported Date**: 2014-06-14T06:59:27.363-07:00
**Planned For Release**: 
**Original CodePlex ID**: 15

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Thanks. The PFX support is optional.

You should import the PFX file using sn.exe in .NET SDK, (sn -i)

http://msdn.microsoft.com/en-us/library/k5b5tt23(v=vs.110).aspx

Then in Obfuscar configuration you can specify `KeyContainer` option to let Obfuscar use the key container to sign the assemblies. The password should never live in Obfuscar configuration file.

About the XAML issue, can you share the binaries with me via support@lextm.com?

**Original Posted Date**: 2014-06-15T05:32:34.63-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

BTW, there is a new revision available for testing, which contains many XAML related fixes,

https://www.nuget.org/packages/Obfuscar/2.0.1-beta

It might solve the crash, but I recommend you share the assemblies with me if possible for further analysis.

**Original Posted Date**: 2014-06-19T23:06:52.453-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Hello.

Thanks.
I will try the new code soon.
And I will also try to reproduce the error in a simpler EXE, since the error appeared on a quite big project.
Right now I am quite busy with some stuff, that's why it took me so long to respond.

**Original Posted Date**: 2014-06-22T06:10:02.95-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Hi.
I have tried the latest version of the code (taken from here: https://obfuscar.codeplex.com/SourceControl/list/changesets?branch=master).
On a simple WPF EXE everything seems OK, but on the EXE that I have tried above I get the same error:

```
Application: MyApp.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Windows.Markup.XamlParseException
Stack:
   at System.Windows.Markup.WpfXamlLoader.Load(System.Xaml.XamlReader, System.Xaml.IXamlObjectWriterFactory, Boolean, System.Object, System.Xaml.XamlObjectWriterSettings, System.Uri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(System.Xaml.XamlReader, Boolean, System.Object, System.Xaml.Permissions.XamlAccessLevel, System.Uri)
   at System.Windows.Markup.XamlReader.LoadBaml(System.IO.Stream, System.Windows.Markup.ParserContext, System.Object, Boolean)
   at System.Windows.Application.LoadComponent(System.Object, System.Uri)
   at KEL_Building_Manager.App.Main()
```

It throws some handled exceptions about not finding "WPFToolkit.Extended.dll" when debugging, but I don't know:
- Why it can't find it (it is next to the EXE - should I add this to the obfuscation too? I think it is signed).
- If this is a big problem or if it is not important.

**Original Posted Date**: 2014-07-28T12:25:37.47-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

Oups! I have just seen your link above.
I have downloaded the wrong version.
I will try again...

**Original Posted Date**: 2014-07-28T12:26:54.317-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

If you can provide a code repository instead of NuGet package it would be nice.
I don't like NuGet so much. :)
Will this work? :
http://obfuscar.googlecode.com/svn/trunk/ obfuscar-read-only

**Original Posted Date**: 2014-07-28T12:35:15.483-07:00

**Comment by lextudio-support** [comment]
**Copied from CodePlex without authors**:

The code repository can be found at https://github.com/lextm/obfuscar

**Original Posted Date**: 2014-12-24T19:25:16.867-08:00

**Comment by lextudio-support** [comment]
If PFX files are used to sign assemblies, they should be put into KeyContainer. WPF issues are known and support is always limited before BAML support is completed.

**Original Closed Date**: 2014-12-24T19:26:26.257-08:00

**Comment by monty241** [comment]
Solved in dot-blur fork.
